cmd/checkpoint/main.go

```diff
@@ -47,49 +48,73 @@ func main() {
-			// kubelet will stop it.
-			if err := os.Remove(activeManifest(tempPodName)); err != nil {
-				glog.Error(err)
+
+		pwc := getPodsWithCheckpointAnnotation(podList)
+		// special case for handling API server readiness
+		if _, ok := pwc[kubeAPIServer]; ok {
+			// Make sure it's actually running. Sometimes we get that
+			// pod manifest back, but the server is not actually running.
+			if _, err := client.Discovery().ServerVersion(); err != nil {
+				delete(pwc, kubeAPIServer)
 			}
-		case isPodRunning(podList, client, actualPodName, namespace):
-			glog.Infof("actual pod %v found, creating checkpoint pod manifest", actualPodName)
-			// The actual is running. Let's snapshot the pod,
-			// clean it up a bit, and then save it to the ignore path for
-			// later use.
-			checkpointPod := createCheckpointPod(podList, actualPodName, namespace)
-			convertSecretsToVolumeMounts(client, &checkpointPod)
-			writeManifest(checkpointPod, tempPodName)
-			checkpoints[checkpointManifest(tempPodName)] = struct{}{}
-			glog.Infof("finished creating checkpoint pod %v manifest at %s\n", tempPodName, checkpointManifest(tempPodName))
-
-		case checkpointed:
-			glog.Info("no actual pod running, installing checkpoint pod static manifest")
-			b, err := ioutil.ReadFile(checkpointManifest(tempPodName))
-			if err != nil {
-				glog.Error(err)
-			} else {
-				if err := ioutil.WriteFile(activeManifest(tempPodName), b, 0644); err != nil {
+		}
+
+		for name, pods := range pwc {
+			_, checkpointed := checkpoints[checkpointManifest(name)]
+
+			switch {
+			case len(pods) == 1 && !checkpointed:
+				glog.Infof("only actual pod %v found, creating checkpoint pod manifest", name)
+				// The actual is running. Let's snapshot the pod,
+				// clean it up a bit, and then save it to the ignore path for
+				// later use.
+				checkpointPod := createCheckpointPod(pods[0], name, namespace)
+				convertSecretsToVolumeMounts(client, &checkpointPod)
+				writeManifest(checkpointPod, name)
+				checkpoints[checkpointManifest(name)] = struct{}{}
+				glog.Infof("finished creating checkpoint pod %v manifest at %s\n", name, checkpointManifest(name))
+
+			case len(pods) == 2:
+				glog.Infof("both checkpoint and actual %v pods running, removing checkpoint pod", name)
+				// Both the temp and actual pods are running.
+				// Remove the temp manifest from the config dir so that the
+				// kubelet will stop it.
+				if err := os.Remove(activeManifest(name)); err != nil {
 					glog.Error(err)
 				}
 			}
 		}
+
+		// start the checkpoint pod if it has not started yet
+		for checkpoint := range checkpoints {
+			pname := podNameFromManifestPath(checkpoint)
+
+			if _, ok := pwc[pname]; !ok {
+				glog.Info("no actual pod running, installing checkpoint pod static manifest")
+				b, err := ioutil.ReadFile(checkpoint)
+				if err != nil {
+					glog.Error(err)
+					continue
+				}
+
+				// Think: touch activeManifest first? do not write if already exists?
```

Comment:
Looks like we have just been overwriting each time. We may need to test the kubelet behavior (e.g. is it just inotify, or does it check for changes). If an overwrite that contains no changes actually triggers a pod restart, we should be more careful about this. But also  as-is wouldn't be a behavior change, so we could leave that to follow up too.